[Final Review in Progress] P2P extension

arthuro555 opened the pull request:
This is a basic multiplayer extension that uses [peer.js](https://peerjs.com/) (a wrapper arround WebRTC to make it easier to use).
It is very basic but also powerful. It doesn't require a server as it uses P2P.

4ian commented:
Very interesting! Multiplayer is often asked, and is a tricky thing to get right, but maybe a basic extension could help. I'll take a look :)
(We'll label this extension as "experimental" though)

arthuro555 commented:
This time I was really careful with casing, I hope I didn't make any casing mistake again 😬 .

arthuro555 commented:
[test_multiplayer.zip](https://github.com/4ian/GDevelop/files/4852156/test_multiplayer.zip)
Here a test project that I quickly put together to test the extension

4ian commented:
> Here a test project that I quickly put together to test the extension

Nice! I'll take a look. Very useful to have something to test, thanks :)

arthuro555 commented on Extensions/Multiplayer_P2P/JsExtension.js:
```diff
@@ -0,0 +1,186 @@
+    extension
+      .addAction(
+        'SendToOne',
+        _('Send to one connected client'),
```
"Send to a specific connected client"?

4ian replied:
Yep!

arthuro555 commented:
I corrected what you told me to correct, it is ready for re-review/merging/the next step

4ian commented:
Do you have something about:

> Do you know about the expected performance of Peer.js? (is this supposed to be scaling to 10, 100 or 1000 connections at the same time?).

And more generally the expected performance for this? 

I tried to see if running two previews could work, but it seems not, can you check if this is normal? (if not it's a bit of a pity, would have been useful to test locally).

We will need an icon for this, I could come up with one if you have a template/svg.

arthuro555 commented:
For scaling, according to https://stackoverflow.com/a/59000150/10994662, WebRTC can have up to 256 connections, maybe 500 on chrome.

> I tried to see if running two previews could work, but it seems not, can you check if this is normal? (if not it's a bit of a pity, would have been useful to test locally).    

If you are talking about my example project, look at the comment at the end of the event sheet to see how to connect the previews.

> We will need an icon for this, I could come up with one if you have a template/svg.

Maybe as template you can use the Peer JS Icon? I am not sure if it would be legally OK. They have the MIT license and I don't see any special copyright on the logo. Here is it: 
https://images.opencollective.com/peer/4912370/logo/256.png

arthuro555 commented:
I found [a cool CC licensed icon](https://iconify.design/icon-sets/bx/bxs-network-chart.html). The only problem is it is a bit harder to see in darker editor themes.

arthuro555 commented:
Uhm what is missing for merging this extension? It looks ready to me and before I work on my older PRs I would like to know if I still need to work on this one

4ian commented on newIDE/app/src/JsExtensionsLoader/BrowserJsExtensionsLoader.js:
```diff
@@ -73,6 +73,11 @@ const jsExtensions = [
     objectsRenderingServiceModules: {},
   },
+  {
+    name: 'P2P Multiplayer',
```
This should be the same name as in the extension declaration

4ian replied:
Also check everywhere that there is no more mention of `Multiplayer` in the extension. This is too low level to be marketed as "multiplayer", I don't want to induce people to error.

4ian commented:
Could we rename the folder to the same as the extension name in its declaration?

4ian commented:
I think we're close to merging this, I just want to be extra sure that the "interface" (i.e: actions/conditions) are very clear and well designed. Because starting from now, we'll have to maintain this forever (well, actually, to avoid this implicit contract of maintaining this forever, we add a label "experimental" so that if we remove this or break compatibility in the future, we can just say "hey it was experimental so not our fault" - but still we don't want to do this.).

4ian commented on Extensions/Multiplayer_P2P/JsExtension.js:
```diff
@@ -0,0 +1,189 @@
+        _('Event triggered by peer'),
+        _('Triggers once when a connected client sends the event'),
+        _('Event _PARAM0_ received from other client'),
+        _('P2P Multiplayer'),
```
`P2P Multiplayer` => `P2P (experimental)` everywhere

4ian commented on Extensions/Multiplayer_P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,126 @@
+  peer: new Peer(),
+
+  /**
+   * All connected p2p clients accessible by id.
```
`All connected p2p clients, keyed by their id`

4ian commented on Extensions/Multiplayer_P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,126 @@
+/**
```
Can you add `// @ts-check` to this file?

4ian commented on Extensions/Multiplayer_P2P/A_peer.js:
```diff
@@ -0,0 +1,70 @@
+parcelRequire=function(e,r,t,n){var i,o="function"==typeof parcelRequire&&parcelRequire,u="function"==typeof require&&require;function f(t,n){if(!r[t]){if(!e[t]){var i="function"==typeof parcelRequire&&parcelRequire;if(!n&&i)return i(t,!0);if(o)return o(t,!0);if(u&&"string"==typeof t)return u(t);var c=new Error("Cannot find module '"+t+"'");throw c.code="MODULE_NOT_FOUND",c}p.resolve=function(r){ ...
```
Would be good to add a comment at the top of the file stating what is this library name, its license and its version + a link to where to get it.

4ian commented:
Basically, let's do everything that is needed so that a new user can understand and try this without asking a single question on the forum. For this, a user needs:
- an example
- a help page to understand what's going, and which will in particular explain quickly what this extension is good for and what are its limitations (if you want to do a MMO with this, well that won't be enough :))
- proper naming of the extension and descriptions of each action/condition.

Sorry I'm picky but we want an extension as perfect as possible - in the past I made the error of putting half backed, poorly maintained extensions in GDevelop and that had negative consequences. The code is something, but what is around (documentation, examples) is very important too. Let's make this a very high quality extension :)

4ian commented on Extensions/P2P/A_peer.js:
```diff
@@ -0,0 +1,71 @@
+// Peerjs, a MIT licensed WebRTC wrapper for P2P communication (https://peerjs.com/).
```
Can we add the exact version number?

4ian commented on Extensions/P2P/peerjs.d.ts:
```diff
@@ -0,0 +1,208 @@
+// Type definitions for the PeerJS class module
```
Nice 👍

arthuro555 commented:
The example is done if you want to take a look at it :eyes:

4ian commented:
Thanks for the wiki page and for the example! Sorry I've been busy, I'll take a look at the example tomorrow and merge this if everything is alright :)

arthuro555 commented:
Did you have some time to look at the example?

4ian commented:
Not yet! Had little time for GDevelop these days, but I'll take a look :)

4ian commented:
Finally got some time for trying the example! :) I tried to run two previews but had a crash:

![image](https://user-images.githubusercontent.com/1280130/87542612-8580c880-c6a3-11ea-9217-2ad2c7f2027e.png)

To reproduce:

* Run two previews
* Wait for the id to show in one (can we show a "getting id..." or something like that while it's not done?)
* Enter the id in the other preview, press Connect.

Might be because of my poor internet connection? Not sure what's happening but the extension should be robust against this for sure.

arthuro555 commented:
I think I fixed it, can you try again please?

4ian commented:
I'll give it another try :) Can you confirm the two previews are supposed to work?

arthuro555 commented:
🤔 It seems like the broker server is being DDoS'd, that's why it is malfunctioning

arthuro555 commented:
https://streamable.com/v09h9j
This is the expected behavior/usage

4ian commented:
Still having this issue: 
![image](https://user-images.githubusercontent.com/1280130/87561146-2c6d6080-c6b4-11ea-89f4-388b6314998f.png)

Give a check at the whole extension for anything that can be null. A crash is almost the worst thing that can happen in GDevelop, this means that your game is broken, that GDevelop will be considered bad quality and will frustrate users.

> 🤔 It seems like the broker server is being DDoS'd, that's why it is malfunctioning

This means that this extension is implicitly relying on a broker server without telling the user (that could think that peer to peer is entirely without a server). It also means that we're officially giving a feature that relies on someone else server. 

I would need you to clearly state this when making such an extension, because otherwise we risk shipping extension to users without fully understanding the implications. What happens if the server is DDOS'ed, what happens if the server is compromised, and should we give the opportunity to the user to use its own server when initializing the extension? :)

Also from my previous message:

> can we show a "getting id..." or something like that while it's not done?

The example is working in the "happy path", but is not in the "sad path", which means that in case of error, there is nothing shown to the user, and no loading indicator/loader. We should fix this before shipping this.

I know it's **not fun nor interesting**, but that's what is giving the **very high quality bar** that we need for GDevelop: a complete example that is robust against all errors, that is clearly showing what is going on, never let you wonder "is this thing broken?". Remember, users are not supposed to open the developer tools ;) 
Otherwise we'll get users to try this and then complain on the forum/Discord (or worse, on an external website) that this is randomly not working, leading me, you or someone else to spend an evening investigating why connections are not working :)

Wend1go commented:
When adding a [Peer](https://peerjs.com/docs.html#api), is it possible to specify the `host` from within your extension? (haven't tried it yet)
According to the Peer.js website you can host your own peer.js server instance. Having the option to use your own server would overcome the threat of having the "[Cloud Peer Server](https://peerjs.com/peerserver.html)" being DDOSed / taken down by it's developers unexpectedly / stolen by aliens or what ever.
I don't think that it is a good idea to rely on this server as the "default" option. I'd rather add a field "host" where you can specify your own server instance and only connect to the "Cloud Peer Server" if no URL ist specified. "For testing / debugging" so to say.
The question is: "Do the Peer.js developers expect that whole game engines start relying on their (free, donation based) service?" or do they see it as just a playground for testing?

arthuro555 commented:
Ok, I worked some more on it. Here are the changes
1. Let the user chose between data loss mode or not instead of always using it
2. Very basic error handling
3. Example integrates clipboard extension now
4. Added ac condition to know when the extension is ready
5. Added an action to change the broker server

If you want top try it again I recommend ussing a local server the main server seems dead. I'll cahnge the wiki to also heavily encourage to use a local server.

The dataloss mode is how the incoming events are handled: when there is the dataloss mode, only the last event data is memorized and used. When disabled, every single data is remembered to be sure it is all being taken into account. Tho only one value can be processed per frame, so it might slow down.

The dataloss mode may be used for example when setting the player position distantzly: you don't care if you missed a position event, you only care about the latest one (current position).
But if you have a point counter, you may want it to be synchronised and don't want any point to get lost/overwritten

In some way you can compare those with TCP and UDP I guess.

arthuro555 commented:
The only thing missing is a way to handle disconnection.

4ian commented:
Thanks for the explanation about the "dataloss", very clear :) I was at first a bit unsure what it was about, I guess proper documentation is important here.

> I'll cahnge the wiki to also heavily encourage to use a local server.

Ok sounds good :) We can add a warning (see `newIDE/app/src/Hints/index.js`) to heavily recommend reading the help and setting up a server.
Ideally in the future if an extension like this start to get popular amongst advanced users, I can try to set up a service to have a GDevelop broker server.

Thanks for the updates, looks pretty good. Let me know when you want a new full review! :)

arthuro555 commented:
It turns out after some testing that peer js cannot hotswap servers without many issues so I cannot do it through an action like I originally intended. I will have to finish my PR for extension properties to choose the server as property before finishing this one.

4ian commented:
Could you do an action that only create the PeerJS object/connection after it's called with the servers?

arthuro555 commented:
Oh yes right I am stupid I can just force the usage of the connection action and make the PeerJs server the default values
________________________________
From: Florian Rival <notifications@github.com>
Sent: Monday, July 27, 2020 12:07:13 AM
To: 4ian/GDevelop <GDevelop@noreply.github.com>
Cc: Arthur Pacaud <arthur.pacaud@hotmail.fr>; Author <author@noreply.github.com>
Subject: Re: [4ian/GDevelop] Basic p2p extension (#1842)


Could you do an action that only create the PeerJS object/connection after it's called with the servers?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/4ian/GDevelop/pull/1842#issuecomment-664046430>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AETT4LRJEYFDWWTQH6PS44LR5SSJDANCNFSM4OMHT36A>.

arthuro555 commented:
I don't like the fact to require an initialization action but I guess it's the best way until extension properties are added.
I am almost done writing the wiki page, I think the code is ready for you to review again.

arthuro555 commented:
Ok the important parts of the wiki are complete and the code now pass CI. I believe it is ready for review.

4ian commented:
Thanks :) I've updated the fixtures in master - would you be able to revert your changes there and merge/rebase master in your branch?

arthuro555 commented:
Done I think

De : Florian Rival<mailto:notifications@github.com>
Envoyé le :dimanche 2 août 2020 20:06
À : 4ian/GDevelop<mailto:GDevelop@noreply.github.com>
Cc : Arthur Pacaud<mailto:arthur.pacaud@hotmail.fr>; Author<mailto:author@noreply.github.com>
Objet :Re: [4ian/GDevelop] Basic p2p extension (#1842)


Thanks :) I've updated the fixtures in master - would you be able to revert your changes there and merge/rebase master in your branch?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/4ian/GDevelop/pull/1842#issuecomment-667706249>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AETT4LV6E2Y2RAW7I5WEZD3R6WTJDANCNFSM4OMHT36A>.

4ian commented on Extensions/P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,308 @@
+  });
+  // Regularly check for disconnection as the built in way is not reliable.
+  var disconnectChecker = function() {
+    if (connection.peerConnection.connectionState == "failed" || connection.peerConnection.connectionState == "disconnected") {
```
Can you check everywhere that you used `===` instead of `==`? Basically, never ever used == in JS.

arthuro555 replied:
I used it on purpose to check both for undefined or null

4ian replied:
For undefined it's fine then! But for strings/number/anything else, always go for === to avoid expensive typecastings and deoptimizations.

4ian commented on Extensions/P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,308 @@
+gdjs.callbacksRuntimeScenePostEvents.push(function() {
+  for(var i in gdjs.evtTools.p2p.lastEventData) {
+    if(typeof gdjs.evtTools.p2p.lastEventData[i] === "object" && gdjs.evtTools.p2p.lastEventData[i].length > 0)
+      gdjs.evtTools.p2p.lastEventData[i].length--;
```
Let's use `pop` instead, I'm afraid manipulating the length of an array can deoptimize it on some engines.

4ian commented on newIDE/app/resources/examples/online-multiplayer/README.md:
```diff
@@ -0,0 +1 @@
+An showcase of some multiplayer possibilities using the P2P extension.
```
A showcase

4ian commented on newIDE/app/src/Hints/index.js:
```diff
@@ -1,8 +1,13 @@
+export type Hint = {|
+  kind: 'warning' | 'info',
+  message: MessageDescriptor,
+  identifier?: AlertMessageIdentifier,
```
Let's call this `dismissIdentifier` to make it very clear that if set, it's used to dismiss the alert.

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+
+    extension
+      .addStrExpression(
+        'GetError',
```
Should this be called `GetLastError`?

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+
+    extension
+      .addStrExpression(
+        'GetDisconnectedPeer',
```
`GetLastDisconnectedPeer` maybe? What do you think?

arthuro555 replied:
True

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+      .addStrExpression(
+        'GetID',
+        _('Get client ID'),
+        _('Gets the current client ID of the current game instance'),
```
Too much "current": `Get the client ID of the current game instance`

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+    extension
+      .addAction(
+        'UseDefaultBroker',
+        _('Connect to default broker server'),
```
Connect to the

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+        'UseDefaultBroker',
+        _('Connect to default broker server'),
+        _('Connects to the default broker server.'),
+        _('Connect to default broker server'),
```
Same

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+      .addAction(
+        'UseDefaultBroker',
+        _('Connect to default broker server'),
+        _('Connects to the default broker server.'),
```
[...] server. It is still recommended that you use a custom broker server if your game is published, so that you have full control over it.

arthuro555 replied:
There is already the hint for that?

4ian replied:
My bad, perfect then!

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+    extension
+      .addAction(
+        'UseOwnBroker',
+        _('Connect to own broker server'),
```
`Connect to a broker server`

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+      .addAction(
+        'Connect',
+        _('Connect to another client'),
+        _('Connects the current client to another client using its id'),
```
Period.

4ian commented on Extensions/P2P/JsExtension.js:
```diff
@@ -0,0 +1,297 @@
+      .addCondition(
+        'OnError',
+        _('An error occured'),
+        _('Triggers once when an error occurs.'),
```
Give a bit more context to the user: "Triggers once when an error occurs. Use `P2P::GetLastError()` expression to get the content of the error if you want to analyse it or display it to the user."

4ian commented on Extensions/P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,308 @@
+  /**
+   * The last peer that has disconnected.
+   */
+  disconnectedPeer: "",
```
Let's rename this to `lastDisconnectedPeerId` to make obvious it's the last and it's an id that we store?

4ian commented on Extensions/P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,308 @@
+  /**
+   * True if a peer diconnected.
+   */
+  peerDisconnected: false,
```
to show that this is valid only once, I usually like naming this with "just": `peerJustDisconnected`

4ian commented:
With the ability to choose a broker server, I think we're reaching a good state! Anything that you can think of on your side that we should fix before merging this? Maybe proofread the wiki page?

4ian commented on Extensions/P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,308 @@
+ * @see gdjs.evtTools.p2p.ready
+ * @returns {boolean}
+ */
+gdjs.evtTools.p2p.isReady = function() {return gdjs.evtTools.p2p.ready;}
```
Let's run Prettier again on all files ;)

4ian commented on Extensions/P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,308 @@
+}
+
+/**
+ * Use default broker server. NOT RECOMMENDED.
```
Use default broker server. This is not recommended for published games, this server should only be used for quick testing in development.

arthuro555 replied:
I would also not recommend it in development as it can be down and cause confusion on why it isn't working.

4ian commented on Extensions/P2P/B_p2ptools.js:
```diff
@@ -0,0 +1,308 @@
+/**
+ * Connects to a custom broker server.
+ * @param {string} host The host of the broker server
+ * @param {number} port
```
[...] The port of the boker server

arthuro555 commented:
> With the ability to choose a broker server, I think we're reaching a good state! Anything that you can think of on your side that we should fix before merging this? Maybe proofread the wiki page?

I think the wiki should be proof readed, maybe some screenshots could be added to it too, and most importantly I would like to try it in "real conditions" (with a few people in different places). I only tried with 2 people until now.

4ian commented:
I've done a first pass on the wiki :) I think if you can try a game with 4/5 people that would be great to check that it's working - I'm fairly confident PeerJS is robust but who knows, this could reveal a bug in the implementation we have if there are more than a few clients!

4ian commented:
I'll give this another try and check on the wiki and we'll merge this when we're confident this is robust 👍 Thanks for all the work on this so far!

arthuro555 commented:
Uhm I was working on a project to test the extension "in a real context/in real conditions" but unfortunately my PC seems to have died. I won't be able to test for a while. I am sorry to ask for this but I would appreciate if some contributor or gdeveloper could try it out for me (cc @Silver-Streak)
Here is a portable build of GDevelop with the extension that I made before my PC died: https://arthuro555-data.s3.eu-de.cloud-object-storage.appdomain.cloud/GDevelop%205%205.0.0-beta98-p2p.exe
It is 6days old I think it is up to date with the latest commit I made.

arthuro555 commented:
Ok all tests are done, I had 6 different people connected at one moment and everything seems good :)

4ian commented:
Excellent! Thanks for checking that :) I'll merge this tomorrow!

> but unfortunately my PC seems to have died

Hope your PC is alright?

arthuro555 commented:
> Hope your PC is alright?

Ah yes sorry I forgot to say that I managed to fix it yesterday. I used the word died died because I thought it was a hardware issue but it was just a corrupt os update :)

4ian commented:
Merged! Thanks for your work on this, it's a very interesting extension and I look forward to how it can be used to make multiplayer games :) 
If this is going well, we should think about how we can simplify even more multiplayer game creation!

arthuro555 commented:
> how we can simplify even more multiplayer game creation!


I actually had an idea and wrote a whole issue about it but then 1 sentence before finishing my browser crashed :/ I'll write it again later but rn I am busy

Btw be careful you did a normal merge not a squash (I told you after immediately on Discord so that you could have rewritten history without much consequences but you don't check messages there I guess and now some people probably have pulled it)
________________________________
From: Florian Rival <notifications@github.com>
Sent: Wednesday, August 19, 2020 9:06:19 AM
To: 4ian/GDevelop <GDevelop@noreply.github.com>
Cc: Arthur Pacaud <arthur.pacaud@hotmail.fr>; Author <author@noreply.github.com>
Subject: Re: [4ian/GDevelop] [Final Review in Progress] P2P extension (#1842)


Merged! Thanks for your work on this, it's a very interesting extension and I look forward to how it can be used to make multiplayer games :)
If this is going well, we should think about how we can simplify even more multiplayer game creation!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/4ian/GDevelop/pull/1842#issuecomment-675893518>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AETT4LUVSVJ2Z3LEHD4TEFLSBN2WXANCNFSM4OMHT36A>.

4ian commented:
Ah too bad I missed that. Too late now but it's ok, I'll manually remove these from the changelog when generated.
